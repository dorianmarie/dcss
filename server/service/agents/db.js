const { sql, updateQuery, whereClause } = require('../../util/sqlHelpers');
const { query, withClientTransaction } = require('../../util/db');
const format = require('pg-format');

async function setAgent(id, updates) {
  return await withClientTransaction(async client => {
    const result = await client.query(updateQuery('agent', { id }, updates));
    return result.rows[0];
  });
}

async function setAgentSocketConfiguration(id, configuration) {

  if (Object.entries(configuration).length === 0) {
    return null
  }

  const result = await withClientTransaction(async client => {
    await client.query(sql`
      DELETE FROM agent_socket_configuration
      WHERE agent_id = ${id};
    `);

    const records = Object.entries(configuration).map(([key, value]) => [
      id,
      key,
      value
    ]);

    const formattedQuery = format(
      `
      INSERT INTO agent_socket_configuration
        (agent_id, key, value)
      VALUES %L
      RETURNING *;
    `,
      records
    );

    const result = await client.query(formattedQuery);

    return result.rows[0] || null;
  });

  if (!result) {
    throw new Error(`Could not set agent configuration`);
  }

  return result;
}

async function setAgentConfiguration(id, configuration) {

  if (Object.entries(configuration).length === 0) {
    return null
  }

  const result = await withClientTransaction(async client => {
    await client.query(sql`
      DELETE FROM agent_configuration
      WHERE agent_id = ${id};
    `);

    const records = Object.entries(configuration).map(([key, value]) => [
      id,
      key,
      value
    ]);

    const formattedQuery = format(
      `
      INSERT INTO agent_configuration
        (agent_id, key, value)
      VALUES %L
      RETURNING *;
    `,
      records
    );

    const result = await client.query(formattedQuery);

    return result.rows[0] || null;
  });

  if (!result) {
    throw new Error(`Could not set agent configuration`);
  }

  return result;
}

async function setAgentInteraction(id, interaction) {
  const result = await withClientTransaction(async client => {
    await client.query(sql`
      DELETE FROM agent_interaction
      WHERE agent_id = ${id};
    `);

    const result = await client.query(sql`
      INSERT INTO agent_interaction
        (agent_id, interaction_id)
      VALUES
        (${id}, ${interaction.id})
      RETURNING *;
    `);

    return result.rows[0] || null;
  });

  if (!result) {
    throw new Error(`Could not set agent interaction`);
  }

  return result;
}

async function createAgent(params) {
  const {
    title,
    name = '',
    description,
    endpoint,
    owner,
    interaction,
    configuration,
    socket
  } = params;

  const agentCreated = await withClientTransaction(async client => {
    const result = await client.query(sql`
      INSERT INTO agent
        (title, name, description, endpoint, owner_id)
      VALUES
        (${title}, ${name}, ${description}, ${endpoint}, ${owner.id})
      RETURNING *
    `);
    return result.rows[0] || null;
  });

  if (!agentCreated) {
    throw new Error(`Could not create agent`);
  }

  await setAgentSocketConfiguration(agentCreated.id, socket);
  await setAgentConfiguration(agentCreated.id, configuration);
  await setAgentInteraction(agentCreated.id, interaction);

  const agent = await query(sql`
    SELECT *
    FROM agent_view
    WHERE id = ${agentCreated.id}
  `);

  return agent.rows[0] || null;
}

async function getAgent(id) {
  const result = await query(sql`
    SELECT * FROM agent_view WHERE id = ${id}
  `);
  return result.rows[0] || null;
}

async function getAgents(where) {
  const result = await query(`
    SELECT *
    FROM agent_view
    ${whereClause(where)}
  `);
  return result.rows;
}

async function getInteractions(id) {
  const result = await query(sql`
    SELECT *
    FROM interaction
    WHERE deleted_at IS NULL
  `);
  return result.rows;
}

exports.createAgent = createAgent;
exports.getAgent = getAgent;
exports.getAgents = getAgents;
exports.setAgent = setAgent;
exports.setAgentSocketConfiguration = setAgentSocketConfiguration;
exports.setAgentConfiguration = setAgentConfiguration;
exports.setAgentInteraction = setAgentInteraction;
exports.getInteractions = getInteractions;
