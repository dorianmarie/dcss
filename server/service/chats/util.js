const { asyncMiddleware } = require('../../util/api');
const { query } = require('../../util/db');
const db = require('./db');
const authdb = require('../session/db');
const cohortdb = require('../cohorts/db');
const notificationdb = require('../notifications/db');
const scenariodb = require('../scenarios/db');

exports.makeChatInviteNotification = async invite => {
  const host = await authdb.getUserById(invite.sender_id);
  const chat = await db.getChat(invite.chat_id);

  if (!chat.scenario_id) {
    const result = await query(sql`
      SELECT scenario_id
      FROM scenario_persona
      WHERE scenario_persona = ${invite.persona_id}
    `);

    chat.scenario_id = result.rows.length ? result.rows[0].scenario_id : null;
  }

  const scenario = chat.scenario_id
    ? await scenariodb.getScenarioById(chat.scenario_id)
    : null;
  const cohort = chat.cohort_id
    ? await cohortdb.getCohortById(chat.cohort_id)
    : null;

  const hostname = host.personalname || host.username;
  const persona = scenario
    ? scenario.personas.find(persona => persona.id === invite.persona_id)
    : null;

  const asRole = persona ? `, as <strong>${persona.name}</strong>` : '';
  const inScenario = scenario ? `the scenario <strong>${scenario.title}</strong>${asRole}` : '';
  const place = cohort
    ? `${inScenario}, which is part of the cohort <strong>${cohort.name}</strong>`
    : inScenario;

  const joinIn = scenario ? ` in ${place}` : '';
  const html = `<strong>${hostname}</strong> has invited you to join them${joinIn}.`;

  return {
    start_at: invite.created_at,
    expire_at: invite.expire_at,
    props: {
      title: 'New invitation!',
      html,
      className: '',
      color: '',
      icon: 'mail',
      time: 0
    },
    rules: { 'session.isLoggedIn': true, 'user.id': invite.receiver_id },
    type: 'invite',
    invite
  };

  //
  // PREVIOUSLY:
  //
  // notificationdb.createNotification({
  //   start_at: data.created_at,
  //   expire_at: data.expire_at,
  //   props: {
  //     title: 'Invitation!',
  //     message:
  //       '{host} has invited you to join them in {scenario}. Follow this {link}.',
  //     className: '',
  //     type: 'info' ,
  //     color: 'pink',
  //     icon: '',
  //     size: 'small',
  //     time: 0
  //   },
  //   rules: { 'session.isLoggedIn': true, 'user.id': data.receiver_id },
  //   type: 'toast'
  // });
};
