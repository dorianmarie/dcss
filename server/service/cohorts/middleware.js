const { asyncMiddleware } = require('../../util/api');
const { getUserRoles } = require('../session/db');
const { getCohortUserRoles } = require('./db');

// const rolesMap = new WeakMap();

exports.requireCohortUserRole = roles => [
  asyncMiddleware(async (req, res, next) => {
    if (!Array.isArray(roles)) {
      roles = [roles];
    }
    const { roles: siteUserRoles } = await getUserRoles(req.session.user.id);

    const { roles: cohortUserRoles } = await getCohortUserRoles(
      req.body.cohort_id,
      req.session.user.id
    );

    if (
      // super admin always gets permission
      siteUserRoles.includes('super_admin') ||
      // owner always gets permission
      cohortUserRoles.includes('owner')
    ) {
      return next();
    }

    // If I have this role both site-wide AND in this cohort,
    // then I have permission to give someone else this role.
    //
    // Q: should "researcher" be allowed to do this?
    //
    if (
      roles.some(role => siteUserRoles.includes(role)) &&
      roles.some(role => cohortUserRoles.includes(role))
    ) {
      return next();
    }

    const error = new Error('Access Denied');
    error.status = 401;
    throw error;
  })
];
