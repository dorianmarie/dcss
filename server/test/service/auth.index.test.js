import Crypto from 'crypto-js';
import { request } from '../';
import { asyncMiddleware } from '../../util/api';

import * as db from '../../service/session/db';
import * as mw from '../../service/session/middleware';

import * as Sendgrid from '@sendgrid/mail';
jest.mock('@sendgrid/mail', () => ({
  setApiKey: jest.fn(),
  send: jest.fn()
}));

import Nodemailer from 'nodemailer';
jest.mock('nodemailer', () => ({
  createTransport: jest.fn()
}));

const SESSION_SECRET = process.env.SESSION_SECRET || 'mit tsl teacher moments';
const error = new Error('something unexpected happened');

let defaultSuperUser = {
  id: 999,
  email: 'super@email.com',
  roles: ['participant', 'super_admin', 'facilitator', 'researcher'],
  is_super: true,
  username: 'superuser',
  is_anonymous: false,
  personalname: 'Super User'
};

jest.mock('../../service/session/db', () => {
  return {
    ...jest.requireActual('../../service/session/db'),
    createUser: jest.fn(),
    getUserById: jest.fn(),
    getUserByProps: jest.fn(),
    updateUser: jest.fn(),
    updateUserWhere: jest.fn()
  };
});

// This cannot be used inside the mock defined below
let amw = jest.requireActual('../../service/session/middleware');

jest.mock('../../service/session/middleware', () => {
  const amw = jest.requireActual('../../service/session/middleware');
  return {
    ...amw,
    createUser: jest.fn(),
    checkForDuplicate: jest.fn(),
    loginUser: jest.fn(),
    requireUser: jest.fn(),
    respondWithUser: jest.fn(),
    refreshSession: jest.fn(),
    resetUserPassword: jest.fn(),
    updateUser: jest.fn()
  };
});

function mockRespondWithUser(user) {
  mw.respondWithUser.mockImplementation(
    asyncMiddleware(async (req, res) => {
      if (!user) {
        return res.send({ user: req.session.user });
      }
      return res.send({ user });
    })
  );
}

jest.mock('password-generator', () => () => 'GENERATED PASSWORD');

describe('/api/session/*', () => {
  let user;

  afterAll(() => {
    jest.restoreAllMocks();
  });

  beforeEach(() => {
    mockRespondWithUser();

    mw.requireUser.mockImplementation(
      asyncMiddleware(async (req, res, next) => {
        next();
      })
    );
    process.env.DCSS_BRAND_NAME_TITLE = 'BRAND X';
  });

  afterEach(() => {
    jest.resetAllMocks();
    delete process.env.DCSS_BRAND_NAME_TITLE;
    delete process.env.SENDGRID_API_KEY;
  });

  describe('GET /api/session/', () => {
    const path = '/api/session/';

    test('success', async () => {
      mockRespondWithUser(defaultSuperUser);

      const response = await request({ path });
      expect(await response.json()).toMatchInlineSnapshot(`
        Object {
          "user": Object {
            "email": "super@email.com",
            "id": 999,
            "is_anonymous": false,
            "is_super": true,
            "personalname": "Super User",
            "roles": Array [
              "participant",
              "super_admin",
              "facilitator",
              "researcher",
            ],
            "username": "superuser",
          },
        }
      `);
      expect(mw.requireUser.mock.calls.length).toBe(1);
      expect(mw.respondWithUser.mock.calls.length).toBe(1);
    });

    test('failure', async () => {
      let count = 0;
      mw.requireUser.mockImplementation(
        asyncMiddleware(async (req, res, next) => {
          count++;
          amw.requireUser(req, res, next);
          count++;
        })
      );

      const response = await request({ path, status: 401 });
      expect(await response.json()).toMatchInlineSnapshot(`
        Object {
          "error": true,
          "message": "User is not authenticated.",
        }
      `);
      expect(mw.requireUser.mock.calls.length).toBe(1);
      expect(mw.respondWithUser.mock.calls.length).toBe(0);
      expect(count).toBe(1);
    });
  });

  describe('/api/session/session', () => {
    const path = '/api/session/session';
    beforeEach(() => {
      mw.requireUser.mockImplementation(
        asyncMiddleware(async (req, res, next) => {
          req.session.user = defaultSuperUser;
          next();
        })
      );
      mw.respondWithUser.mockImplementation(amw.respondWithUser);
    });

    test('success', async () => {
      mw.refreshSession.mockImplementation(amw.refreshSession);
      db.getUserById.mockImplementation(async () => defaultSuperUser);
      const response = await request({ path });
      expect(await response.json()).toMatchInlineSnapshot(`
        Object {
          "user": Object {
            "email": "super@email.com",
            "id": 999,
            "is_anonymous": false,
            "is_super": true,
            "personalname": "Super User",
            "roles": Array [
              "participant",
              "super_admin",
              "facilitator",
              "researcher",
            ],
            "username": "superuser",
          },
        }
      `);
    });

    test('failure', async () => {
      let count = 0;
      mw.refreshSession.mockImplementation(
        asyncMiddleware(async (req, res, next) => {
          count++;
          amw.refreshSession(req, res, next);
          count++;
        })
      );
      db.getUserById.mockImplementation(async () => {
        return null;
      });
      const response = await request({ path, status: 404 });
      expect(await response.json()).toMatchInlineSnapshot(`
        Object {
          "error": true,
          "message": "User does not exist.",
        }
      `);
      expect(count).toBe(2);
    });
  });

  describe('/api/session/logout', () => {
    test('success', async () => {
      const response = await request({
        path: '/api/session/logout',
        method: 'post'
      });
      expect(await response.json()).toMatchInlineSnapshot(`
        Object {
          "logout": true,
        }
      `);
    });
  });

  // router.post('/login', [
  //   validateRequestBody,
  //   validateRequestUsernameAndEmail,
  //   loginUser,
  //   respondWithUser
  // ]);
  // Crypto.AES.encrypt(password, SESSION_SECRET).toString();
  //
  describe('/api/session/login', () => {
    const method = 'post';
    const path = '/api/session/login';
    const salt = 'b5a6b7c530735eba';
    const hash =
      '15e33cafdcbecf8b43c12887897411337c7085c80994cecbb36f91188aac9290e6004d1bc5e0bcd25e471f48a5f450458998bb02e663ea4963702623eb58f5d6';

    beforeEach(() => {
      mockRespondWithUser(defaultSuperUser);
      db.updateUser.mockImplementation(async () => {});
      db.getUserById.mockImplementation(async () => defaultSuperUser);
    });

    describe('success', () => {
      test('single_use_password: false', async () => {
        mw.loginUser.mockImplementation(amw.loginUser);

        const single_use_password = false;

        db.getUserByProps.mockImplementation(async () => ({
          ...defaultSuperUser,
          salt,
          hash,
          single_use_password
        }));

        const password = Crypto.AES.encrypt(
          'bp7UJZQzIL5Un49xuyKV',
          SESSION_SECRET
        ).toString();
        const body = {
          username: 'superuser',
          password
        };

        const response = await request({ path, body, method });
        expect(await response.json()).toMatchInlineSnapshot(`
          Object {
            "user": Object {
              "email": "super@email.com",
              "id": 999,
              "is_anonymous": false,
              "is_super": true,
              "personalname": "Super User",
              "roles": Array [
                "participant",
                "super_admin",
                "facilitator",
                "researcher",
              ],
              "username": "superuser",
            },
          }
        `);
        expect(db.getUserById.mock.calls.length).toBe(1);
        expect(db.updateUser.mock.calls.length).toBe(0);
      });

      test('single_use_password: true', async () => {
        mw.loginUser.mockImplementation(amw.loginUser);

        const single_use_password = true;

        db.getUserByProps.mockImplementation(async () => ({
          ...defaultSuperUser,
          salt,
          hash,
          single_use_password
        }));

        const password = Crypto.AES.encrypt(
          'bp7UJZQzIL5Un49xuyKV',
          SESSION_SECRET
        ).toString();
        const body = {
          username: 'superuser',
          password
        };

        const response = await request({ path, body, method });
        expect(await response.json()).toMatchInlineSnapshot(`
          Object {
            "user": Object {
              "email": "super@email.com",
              "id": 999,
              "is_anonymous": false,
              "is_super": true,
              "personalname": "Super User",
              "roles": Array [
                "participant",
                "super_admin",
                "facilitator",
                "researcher",
              ],
              "username": "superuser",
            },
          }
        `);
        expect(db.getUserById.mock.calls.length).toBe(1);
        expect(db.updateUser.mock.calls.length).toBe(1);
      });
    });

    describe('failure', () => {
      let status = 401;

      test('This account requires a password', async () => {
        let count = 0;
        mw.loginUser.mockImplementation(
          asyncMiddleware(async (req, res, next) => {
            count++;
            amw.loginUser(req, res, next);
            count++;
          })
        );

        const single_use_password = false;

        db.getUserByProps.mockImplementation(async () => ({
          ...defaultSuperUser,
          salt,
          hash,
          single_use_password
        }));

        const body = {
          username: 'superuser'
        };

        const response = await request({ path, body, method, status });
        expect(await response.json()).toMatchInlineSnapshot(`
          Object {
            "error": true,
            "message": "This account requires a password.",
          }
        `);
        expect(db.getUserById.mock.calls.length).toBe(0);
        expect(db.updateUser.mock.calls.length).toBe(0);
        expect(count).toBe(2);
      });

      test('This account does not require a password', async () => {
        let count = 0;
        mw.loginUser.mockImplementation(
          asyncMiddleware(async (req, res, next) => {
            count++;
            amw.loginUser(req, res, next);
            count++;
          })
        );

        const single_use_password = false;

        db.getUserByProps.mockImplementation(async () => ({
          ...defaultSuperUser,
          single_use_password
        }));

        const password = Crypto.AES.encrypt(
          'bp7UJZQzIL5Un49xuyKV',
          SESSION_SECRET
        ).toString();
        const body = {
          username: 'superuser',
          password
        };

        const response = await request({ path, body, method, status });
        expect(await response.json()).toMatchInlineSnapshot(`
          Object {
            "error": true,
            "message": "This account does not require a password.",
          }
        `);
        expect(db.getUserById.mock.calls.length).toBe(0);
        expect(db.updateUser.mock.calls.length).toBe(0);
        expect(count).toBe(2);
      });

      test('Invalid password', async () => {
        let count = 0;
        mw.loginUser.mockImplementation(
          asyncMiddleware(async (req, res, next) => {
            count++;
            amw.loginUser(req, res, next);
            count++;
          })
        );

        const single_use_password = false;

        db.getUserByProps.mockImplementation(async () => ({
          ...defaultSuperUser,
          single_use_password
        }));

        const password = Crypto.AES.encrypt('nope', SESSION_SECRET).toString();
        const body = {
          username: 'superuser',
          password
        };

        const response = await request({ path, body, method, status });
        expect(await response.json()).toMatchInlineSnapshot(`
          Object {
            "error": true,
            "message": "This account does not require a password.",
          }
        `);
        expect(db.getUserById.mock.calls.length).toBe(0);
        expect(db.updateUser.mock.calls.length).toBe(0);
        expect(count).toBe(2);
      });
    });
  });

  describe('POST /api/session/', () => {
    const method = 'post';
    const path = '/api/session/';

    describe('success', () => {
      test('with password', async () => {
        mw.createUser.mockImplementation(amw.createUser);
        mw.checkForDuplicate.mockImplementation((req, res, next) => next());
        db.getUserByProps.mockImplementation(() => false);
        db.createUser.mockImplementation(() => defaultSuperUser);

        const password = Crypto.AES.encrypt(
          'bp7UJZQzIL5Un49xuyKV',
          SESSION_SECRET
        ).toString();

        const body = {
          username: 'superuser',
          password
        };

        const response = await request({ path, method, body });
        expect(await response.json()).toMatchInlineSnapshot(`
          Object {
            "user": Object {
              "email": "super@email.com",
              "id": 999,
              "is_anonymous": false,
              "is_super": true,
              "personalname": "Super User",
              "roles": Array [
                "participant",
                "super_admin",
                "facilitator",
                "researcher",
              ],
              "username": "superuser",
            },
          }
        `);
        expect(mw.checkForDuplicate.mock.calls.length).toBe(1);
        expect(mw.respondWithUser.mock.calls.length).toBe(1);
      });

      test('without password (anonymous users)', async () => {
        mw.createUser.mockImplementation(amw.createUser);
        mw.checkForDuplicate.mockImplementation((req, res, next) => next());
        db.getUserByProps.mockImplementation(() => false);
        db.createUser.mockImplementation(() => defaultSuperUser);

        const password = Crypto.AES.encrypt(
          'bp7UJZQzIL5Un49xuyKV',
          SESSION_SECRET
        ).toString();

        const body = {
          username: 'superuser'
        };

        const response = await request({ path, method, body });
        expect(await response.json()).toMatchInlineSnapshot(`
          Object {
            "user": Object {
              "email": "super@email.com",
              "id": 999,
              "is_anonymous": false,
              "is_super": true,
              "personalname": "Super User",
              "roles": Array [
                "participant",
                "super_admin",
                "facilitator",
                "researcher",
              ],
              "username": "superuser",
            },
          }
        `);
        expect(mw.checkForDuplicate.mock.calls.length).toBe(1);
        expect(mw.respondWithUser.mock.calls.length).toBe(1);
      });
    });

    describe('failure', () => {
      test('User could not be created.', async () => {
        mw.createUser.mockImplementation(amw.createUser);
        mw.checkForDuplicate.mockImplementation((req, res, next) => next());
        db.getUserByProps.mockImplementation(() => false);
        db.createUser.mockImplementation(() => false);

        const password = Crypto.AES.encrypt(
          'bp7UJZQzIL5Un49xuyKV',
          SESSION_SECRET
        ).toString();

        const body = {
          username: 'superuser',
          password
        };

        const response = await request({ path, method, body, status: 500 });
        expect(await response.json()).toMatchInlineSnapshot(`
          Object {
            "error": true,
            "message": "User could not be created.",
          }
        `);
        expect(mw.checkForDuplicate.mock.calls.length).toBe(1);
        expect(mw.respondWithUser.mock.calls.length).toBe(0);
      });
    });
  });

  describe('/api/session/usernames/:username/exists', () => {
    describe('success', () => {
      test('does not exist', async () => {
        mw.checkForDuplicate.mockImplementation(amw.checkForDuplicate);
        db.getUserByProps.mockImplementation(() => false);

        const response = await request({
          path: `/api/session/usernames/whatever/exists`
        });
        expect(await response.json()).toMatchInlineSnapshot(`
          Object {
            "username": "whatever",
          }
        `);
        expect(mw.checkForDuplicate.mock.calls.length).toBe(1);
        expect(db.getUserByProps.mock.calls.length).toBe(1);
      });
    });

    describe('failure', () => {
      test('Username is already in use', async () => {
        mw.checkForDuplicate.mockImplementation(amw.checkForDuplicate);
        db.getUserByProps.mockImplementation(() => true);

        const response = await request({
          path: `/api/session/usernames/whatever/exists`,
          status: 409
        });
        expect(await response.json()).toMatchInlineSnapshot(`
          Object {
            "error": true,
            "message": "Username is already in use.",
          }
        `);
        expect(mw.checkForDuplicate.mock.calls.length).toBe(1);
        expect(db.getUserByProps.mock.calls.length).toBe(1);
      });
    });
  });

  describe('/api/session/reset', () => {
    const path = '/api/session/reset';

    describe('success', () => {
      test('success', async () => {
        mw.resetUserPassword.mockImplementation(async (req, res) => {
          res.json({ reset: true, reason: 'Success' });
        });

        const body = {
          email: 'super@email.com'
        };
        const response = await request({ path, body, method: 'post' });
        expect(await response.json()).toMatchInlineSnapshot(`
          Object {
            "reason": "Success",
            "reset": true,
          }
        `);
      });
      test('Nodemailer', async () => {
        mw.resetUserPassword.mockImplementation(amw.resetUserPassword);

        let calls = 0;
        db.getUserByProps.mockImplementation(() => {
          if (!calls) {
            calls++;
            return false;
          }
          return defaultSuperUser;
        });

        const sendMail = jest.fn();

        Nodemailer.createTransport.mockImplementation(
          jest.fn(() => {
            return {
              sendMail
            };
          })
        );

        const body = {
          email: 'super@email.com'
        };

        const response = await request({ path, body, method: 'post' });
        expect(await response.json()).toMatchInlineSnapshot(`
          Object {
            "reason": "Success",
            "reset": true,
          }
        `);

        expect(Nodemailer.createTransport.mock.calls[0][0]).toMatchSnapshot({
          auth: {
            pass: expect.any(String),
            user: expect.any(String)
          }
        });

        expect(sendMail.mock.calls[0]).toMatchInlineSnapshot(`
          Array [
            Object {
              "from": "BRAND X <teachermoments@mit.edu>",
              "html": "<p>You are receiving this email because a request was made to reset your BRAND X password. The following password may only be used once. After you've logged in, go to Settings and update your password.</p>
          <p>
          Single-use password: <code>GENERATED PASSWORD</code>
          </p>

          <p>
          <a href=\\"undefined/settings\\">Click here to update your account settings: undefined/settings</a>
          </p>

          <p>
          <small>
          Massachusetts Institute of Technology NE49<br>
          600 Technology Square<br>
          Cambridge, MA 02139
          </small>
          </p>
          ",
              "subject": "BRAND X Single-use password request",
              "text": "
          You are receiving this email because a request was made to reset your BRAND X password.
          The following password may only be used once. After you've logged in, go to Settings and update your password.



          Single-use password: GENERATED PASSWORD



          Click here to update your account settings: undefined/settings


          ------------------------------------------
          Massachusetts Institute of Technology NE49
          600 Technology Square
          Cambridge, MA 02139
          ",
              "to": "",
            },
          ]
        `);
      });

      test('Sendgrid', async () => {
        process.env.SENDGRID_API_KEY = 'x';
        mw.resetUserPassword.mockImplementation(amw.resetUserPassword);

        let calls = 0;
        db.getUserByProps.mockImplementation(() => {
          if (!calls) {
            calls++;
            return false;
          }
          return defaultSuperUser;
        });

        const body = {
          email: 'super@email.com'
        };

        const response = await request({ path, body, method: 'post' });
        expect(await response.json()).toMatchInlineSnapshot(`
          Object {
            "reason": "Success",
            "reset": true,
          }
        `);

        expect(Sendgrid.send.mock.calls[0]).toMatchInlineSnapshot(`
          Array [
            Object {
              "from": "BRAND X <teachermoments@mit.edu>",
              "html": "<p>You are receiving this email because a request was made to reset your BRAND X password. The following password may only be used once. After you've logged in, go to Settings and update your password.</p>
          <p>
          Single-use password: <code>GENERATED PASSWORD</code>
          </p>

          <p>
          <a href=\\"undefined/settings\\">Click here to update your account settings: undefined/settings</a>
          </p>

          <p>
          <small>
          Massachusetts Institute of Technology NE49<br>
          600 Technology Square<br>
          Cambridge, MA 02139
          </small>
          </p>
          ",
              "subject": "BRAND X Single-use password request",
              "text": "
          You are receiving this email because a request was made to reset your BRAND X password.
          The following password may only be used once. After you've logged in, go to Settings and update your password.



          Single-use password: GENERATED PASSWORD



          Click here to update your account settings: undefined/settings


          ------------------------------------------
          Massachusetts Institute of Technology NE49
          600 Technology Square
          Cambridge, MA 02139
          ",
              "to": "",
            },
          ]
        `);
      });
    });

    describe('failure', () => {
      test('Raw email', async () => {
        mw.resetUserPassword.mockImplementation(amw.resetUserPassword);

        db.getUserByProps.mockImplementation(() => {
          return defaultSuperUser;
        });

        const body = {
          email: 'super@email.com'
        };

        const response = await request({ path, body, method: 'post' });
        expect(await response.json()).toMatchInlineSnapshot(`
          Object {
            "reason": "Cannot use a plain text email address to generate a password reset request",
            "reset": false,
          }
        `);

        expect(db.getUserByProps.mock.calls.length).toBe(1);
      });

      test('Nodemailer', async () => {
        mw.resetUserPassword.mockImplementation(amw.resetUserPassword);

        let calls = 0;
        db.getUserByProps.mockImplementation(() => {
          if (!calls) {
            calls++;
            return false;
          }
          return defaultSuperUser;
        });

        const sendMail = jest.fn();

        Nodemailer.createTransport.mockImplementation(
          jest.fn(() => {
            return {
              sendMail() {
                throw error;
              }
            };
          })
        );

        const body = {
          email: 'super@email.com'
        };

        const response = await request({
          path,
          body,
          method: 'post',
          status: 401
        });
        expect(await response.json()).toMatchInlineSnapshot(`
          Object {
            "error": true,
            "message": "something unexpected happened",
          }
        `);

        expect(Nodemailer.createTransport.mock.calls[0][0]).toMatchSnapshot({
          auth: {
            pass: expect.any(String),
            user: expect.any(String)
          }
        });

        expect(sendMail.mock.calls.length).toBe(0);
      });

      test('Sendgrid', async () => {
        process.env.SENDGRID_API_KEY = 'x';
        mw.resetUserPassword.mockImplementation(amw.resetUserPassword);

        let calls = 0;
        db.getUserByProps.mockImplementation(() => {
          if (!calls) {
            calls++;
            return false;
          }
          return defaultSuperUser;
        });

        const body = {
          email: 'super@email.com'
        };

        Sendgrid.send.mockImplementation(() => {
          throw error;
        });

        const response = await request({
          path,
          body,
          method: 'post',
          status: 401
        });
        expect(await response.json()).toMatchInlineSnapshot(`
          Object {
            "error": true,
            "message": "something unexpected happened",
          }
        `);

        expect(Sendgrid.send.mock.calls[0]).toMatchInlineSnapshot(`
          Array [
            Object {
              "from": "BRAND X <teachermoments@mit.edu>",
              "html": "<p>You are receiving this email because a request was made to reset your BRAND X password. The following password may only be used once. After you've logged in, go to Settings and update your password.</p>
          <p>
          Single-use password: <code>GENERATED PASSWORD</code>
          </p>

          <p>
          <a href=\\"undefined/settings\\">Click here to update your account settings: undefined/settings</a>
          </p>

          <p>
          <small>
          Massachusetts Institute of Technology NE49<br>
          600 Technology Square<br>
          Cambridge, MA 02139
          </small>
          </p>
          ",
              "subject": "BRAND X Single-use password request",
              "text": "
          You are receiving this email because a request was made to reset your BRAND X password.
          The following password may only be used once. After you've logged in, go to Settings and update your password.



          Single-use password: GENERATED PASSWORD



          Click here to update your account settings: undefined/settings


          ------------------------------------------
          Massachusetts Institute of Technology NE49
          600 Technology Square
          Cambridge, MA 02139
          ",
              "to": "",
            },
          ]
        `);
      });

      test('no user', async () => {
        process.env.SENDGRID_API_KEY = 'x';
        mw.resetUserPassword.mockImplementation(amw.resetUserPassword);

        let calls = 0;
        db.getUserByProps.mockImplementation(() => {
          if (!calls) {
            calls++;
            return false;
          }
          return null;
        });

        const body = {
          email: 'super@email.com'
        };

        const response = await request({
          path,
          body,
          method: 'post',
          status: 401
        });

        expect(await response.json()).toMatchInlineSnapshot(`
          Object {
            "error": true,
            "message": "The email provided does not match any existing account.",
          }
        `);
      });

      test('failure', async () => {
        mw.resetUserPassword.mockImplementation(async (req, res) => {
          res.json({
            error: true,
            reason: 'The email provided does not match any existing account.'
          });
        });

        const body = {
          email: 'super@email.com'
        };
        const response = await request({ path, body, method: 'post' });
        expect(await response.json()).toMatchInlineSnapshot(`
          Object {
            "error": true,
            "reason": "The email provided does not match any existing account.",
          }
        `);
      });
    });
  });

  describe('PUT /api/session/', () => {
    const method = 'put';
    const path = '/api/session/';

    beforeEach(() => {
      mw.requireUser.mockImplementation(
        asyncMiddleware(async (req, res, next) => {
          req.session.user = defaultSuperUser;
          next();
        })
      );
    });

    test('with password', async () => {
      const body = {
        email: 'another@email.com',
        personalname: 'whatever',
        username: 'something-else'
      };

      mw.updateUser.mockImplementation(amw.updateUser);
      mw.checkForDuplicate.mockImplementation((req, res, next) => next());
      db.getUserByProps.mockImplementation(() => false);
      db.updateUser.mockImplementation((id, updates) => {
        return {
          ...defaultSuperUser,
          ...updates
        };
      });

      const password = Crypto.AES.encrypt(
        'bp7UJZQzIL5Un49xuyKV',
        SESSION_SECRET
      ).toString();

      const response = await request({ path, method, body });
      expect(await response.json()).toMatchInlineSnapshot(`
        Object {
          "user": Object {
            "email": "another@email.com",
            "id": 999,
            "is_anonymous": false,
            "is_super": true,
            "personalname": "whatever",
            "roles": Array [
              "participant",
              "super_admin",
              "facilitator",
              "researcher",
            ],
            "username": "something-else",
          },
        }
      `);
      expect(mw.checkForDuplicate.mock.calls.length).toBe(1);
      expect(mw.respondWithUser.mock.calls.length).toBe(1);
      expect(db.updateUser.mock.calls[0]).toMatchInlineSnapshot(`
        Array [
          999,
          Object {
            "email": "another@email.com",
            "personalname": "whatever",
            "username": "something-else",
          },
        ]
      `);
    });

    test('failure', async () => {
      mw.updateUser.mockImplementation(amw.updateUser);
      mw.checkForDuplicate.mockImplementation((req, res, next) => next());
      db.getUserByProps.mockImplementation(() => false);
      db.updateUser.mockImplementation(() => false);

      const password = Crypto.AES.encrypt(
        'bp7UJZQzIL5Un49xuyKV',
        SESSION_SECRET
      ).toString();

      const body = {
        username: 'superuser',
        password
      };

      const response = await request({ path, method, body, status: 500 });
      expect(await response.json()).toMatchInlineSnapshot(`
        Object {
          "error": true,
          "message": "User could not be updated.",
        }
      `);
      expect(mw.checkForDuplicate.mock.calls.length).toBe(1);
      expect(mw.respondWithUser.mock.calls.length).toBe(0);
    });
  });
});
