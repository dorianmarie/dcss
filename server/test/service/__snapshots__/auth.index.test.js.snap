// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/api/session/* /api/session/reset failure Nodemailer 2`] = `
Object {
  "auth": Object {
    "pass": Any<String>,
    "user": Any<String>,
  },
  "host": "smtp.mailtrap.io",
  "port": 2525,
}
`;

exports[`/api/session/* /api/session/reset success Nodemailer 2`] = `
Object {
  "auth": Object {
    "pass": Any<String>,
    "user": Any<String>,
  },
  "host": "smtp.mailtrap.io",
  "port": 2525,
}
`;
